### ISSUE TITLE
Memory Leak in Python

### ISSUE THREAD

**Issue by chrisrossi** [grpc/grpc#22851]
The following Python script illustrates the issue:

```python
import ctypes
import gc
import grpc

from google.cloud.client import ClientWithProject
from google.cloud._helpers import make_secure_channel
from google.cloud._http import DEFAULT_USER_AGENT

"""
This code exposes a memory leak involving ``grpc`` and ``google.auth``. Under
normal circumstances this isn't an issue, since a well factored program will
only create one grpc communication channel, but in programs that repeatedly
create communication channels, those channels are not cleaned up properly and
cause a memory leak.

To reproduce this, install `google-cloud-datastore` via `pip` into virtual
environment and then run this script.
"""

client = ClientWithProject()
"""
We only need one client, which will have our credentials in it. Calling it
with no arguments assumes that `GOOGLE_APPLICATION_CREDENTIALS` is set in the
environment. These credentials will need to point to a Google project with an
instance of Datastore or Firestore in Datastore mode.
"""


def show_leak():
    """
    The problem is when we make a channel, there's an instance of
    `grpc._plugin_wrapping._Plugin` getting left behind that isn't cleaned up.
    """
    make_channel()

    """
    We haven't assigned the return value to anything, so the channel has fallen
    out of scope and should be cleaned up already. If we go digging around in
    the objects tracked by the garbage collector, however, we find that there
    is still an instance of `_Plugin`. The garbage collector isn't tracking a
    referrer that would explain why it's still in memory, but its
    PyObject.refcnt is 1, so it hasn't been cleaned up yet.
    """

    print("one")
    show_objects(grpc._plugin_wrapping._Plugin)
    """
    Output:

    one
    _Plugin: 1 instances
        refcnt: 1  referrers: 0
    """

    """
    If we create more channels, the instances just keep ratcheting up.
    """
    make_channel()

    print("two")
    show_objects(grpc._plugin_wrapping._Plugin)
    """
    Output:

    _Plugin: 2 instances
        refcnt: 1  referrers: 0
        refcnt: 1  referrers: 0
    """

    """
    The `grpc` library does make use of some C/cython code, so I supposed it's
    possible that some C code has bypassed normal Python reference counting and
    spoiled the count somehow. This is purely conjecture.

    But we can keep doing this all day.
    """
    make_channel()

    print("three")
    show_objects(grpc._plugin_wrapping._Plugin)
    """
    Output:

    three
    _Plugin: 3 instances
        refcnt: 1  referrers: 0
        refcnt: 1  referrers: 0
        refcnt: 1  referrers: 0
    """


class PyObject(ctypes.Structure):
    """C structure pointing to a PyObject instance.

    This allows us to access the reference count for an object without having
    to hold a reference to it in Python, and thus impacting its reference
    count.
    """
    _fields_ = [("refcnt", ctypes.c_long)]


def make_channel():
    """Create the gRPC communication channel.

    Uses credentials from the client.
    """
    channel = make_secure_channel(
        client._credentials,
        DEFAULT_USER_AGENT,
        "datastore.googleapis.com",
    )
    return channel


def find_objects(cls):
    """Search for objects of given class, `cls`, tracked by garbage collector.

    This function starts with a call `gc.collect()` to get rid of any instances
    that may still be around due to a reference cycle. Theoretically, then, any
    objects found are still around because there are active references to them.

    Returns:
        generator[tuple[PyObject, int]]: A tuple, `(py_object, n_referrers)` for
            each object found, where `py_object` is a `PyObject` reference to
            the object, and `n_referrers` is the number of objects tracked by
            the garbage collector that refer to the found object.
    """
    gc.collect()

    objects = []
    gc_objects = gc.get_objects()
    for o in gc_objects:
        if type(o) is not cls:
            continue

        args = [o]
        referrers = [
            referrer for referrer in gc.get_referrers(*args)
            # exclude referrers created during this function call
            if referrer not in (gc_objects, args)
        ]
        # We should be left only with referrers we didn't create during this
        # function call.
        n_referrers = len(referrers)

        # Create a PyObject reference to the object that we can examine without
        # affecting the Python object's reference count
        py_object = PyObject.from_address(id(o))

        objects.append((py_object, n_referrers))

    return objects


def show_objects(cls):
    """Call `find_objects` and print the results."""
    objects = find_objects(cls)
    gc.collect()  # collect reference cycles made in call to find_objects
    print(f"{cls.__name__}: {len(objects)} instances")
    for py_object, n_referrers in objects:
        print(f"    refcnt: {py_object.refcnt}  referrers: {n_referrers}")


if __name__ == "__main__":
    show_leak()
```

**Comment by lidizheng** [comment]
@chrisrossi Can you help us identify if the leak is same as https://github.com/grpc/grpc/issues/22123? If so, an ongoing PR might be able to fix this leak.

**Comment by chrisrossi** [comment]
@lidizheng I've just tested again using `master` and can confirm the problem has resolved itself, presumably due to #22855. Thank you!

**Comment by lidizheng** [comment]
Thank you for providing the detailed reproduction case. Great way of using `ctypes`! Fixed by #22855, closing this issue.
